### DISCUSSION TITLE
WearOS Support

### DISCUSSION THREAD

**Discussion by caffeinejolt** [organicmaps/organicmaps#3273] [Ideas]
Now that WearOS is becoming more prevalent, it would be pretty cool to offer a companion WearOS app.  A concern with watch use is always going to be battery efficiency.  It would be great to just glance at the watch and have some basic options.

**Comment by biodranik** [comment]
The idea is good, but the battery consumption is not suitable for everyone. There are already some issues related to watches.

**Reply by caffeinejolt** [reply]
Well, it might just be basic support in the beginning - more of a companion app than a full-fledged map/nav system.  Some things which might not impact battery use as much (but still offer a lot of what people want):

Show distance / time remaining on current route
Show an indicator if your are on or have strayed off route
Show current elevation along with elevation difference traveled so far on route

I am not suggesting that we attempt to replicate navigation functionality on a watch, but rather that the watch can be easily checked to get key information by talking to the app running on the phone.

**Reply by biodranik** [reply]
To make it more productive, let's focus on a proper design of such an app. What should be there (for a minimum viable product level)? How it should look? How it should work?

**Reply by grevory** [reply]
Is there a place to contribute ideas?

**Reply by hb0nd** [reply]
> Is there a place to contribute ideas?

The best place is in this issue :)

**Reply by grevory** [reply]
WRT to WearOS, I think a UI which is predominately an arrow which points you in the direction of the trail you are on, and toward waypoints if they exist (I couldn't find a way to add waypoints). This would make trekking easier so you would not need to keep pulling out your phone on a trail. And also a notification if you veer X meters away from trail you have set.

I am doing a hike this summer where there is no trail and game trail can often take you off course. I would love a minimum effort solution to solve that problem :)

**Reply by biodranik** [reply]
What are "waypoints" in this context? Intermediate stops on the way? Select any place after building a route, and press "Add stop" at the bottom.

**Comment by mxdanger** [comment]
It might be even better to add support for both watchOS and Wear OS It would be possible to add full navigation on watchOS if that was the end goal but that might not be possible on Wear OS devices for now. So just having it act solely as a companion seems like the best choice for batter efficiency as you mentioned.

**Comment by ghost** [comment]
+1 for WearOS or HarmonyOS support! I've installed OM on my Huawei Watch 3 and it's a perfect replacement for other map apps, especially for having offline maps on my wrist. If you need UI/UX suggestions (including a working prototype with Figma) let me know :)
Cheers!

**Comment by eamag** [comment]
+1 as pixel watch also uses wear os

**Comment by grevory** [comment]
Thank you! Yes, intermediate planned stops or landmarks.

The arrow could point to the course or to the waypoints.

If a user has veered too far off their course they get buzzed, so they
don't need to keep the map out while traveling.

**Comment by TheWanderer1983** [comment]
+1 Samsung Galaxy Watch 7 uses WearOS
+1 OnePlus 2R watch uses WearOS also.

**Comment by TheWanderer1983** [comment]
Saved/Starred locations become easily accessible.

**Comment by janlr123** [comment]
Is there any update here. With more and more wearOS decides on the marked with linger battery life a watch version would be amazing.

**Comment by alexnagorn** [comment]
Hello. I'v installed organic maps on Oneplus watch 3 46mm. It works really fast. For my feelings faster then Gmaps and better then list of all wear os outdoor applications. I tried to build an outdoor route and I succeeded. However, size of the control elements and absence of the 'Cancel' button ruins all normal use case. As well wheel works just for zooming in any scrolling direction. It takes wrong coordinates (around 500m from actual point) when searching GPS and can't locate North so compas arrow also wrong... I would tell it has huge potential and can easy overtake Garmin and Suunto just needs some UX/UI and pairing of the internal sensors. Hope some developers would see this comment. Regards!

**Comment by wrathgewp** [comment]
+1 Galaxy Watch 7 uses WearOS

There are already some OSM clients for WearOS, but they are all close source with In-app purchases.

**Comment by datend3nker** [comment]
I’ve started a proof-of-concept Wear OS companion at https://github.com/datend3nker/organicmaps-wearos. The current state is mostly “vibe coded” because I’m not familiar with Kotlin (I’m an embedded security engineer). It’s intended to explore UI/UX and find the most useful flow; once I’m happy with the design I’ll rewrite the app from scratch without AI assistance.

I’m at university and short on time, so please be patient  I’ll iterate on this when I can and welcome feedback on the approach and UX.

**Reply by datend3nker** [reply]
app is in some sort of functional state... would appreciate some feedback on the UI to start with the rewrite

**Reply by biodranik** [reply]
Would you mind sharing a video of the demo?

**Reply by datend3nker** [reply]
Video file is too big so here the link to the file: [https://nextcloud.ettner.cloud/s/mGrCsMtS44eje5D](https://nextcloud.ettner.cloud/s/mGrCsMtS44eje5D)
Sry for the wired cut out. My hyprland is not fully configured to work with OBS.
As you can see, navigation from phone to watch works. The navigation arrow orientates with the physical orientation to ease navigation.
Plan is to enable search via watch and maybe enable some sort of stand-alone mode. Currently, it is not completely FOSS as it depends heavily on Google.

Have not tested it on a real device

**Reply by biodranik** [reply]
Thanks, just showing directions from the app on a phone may already be useful, and maybe stopping the navigation. The UI likely can be improved, would be great to see how other apps implemented it (make screenshots to analyze and compare), and make it similar or better in OM. One case that comes to mind: approaching an intermediate stop (it should be visualized differently).

Vibecode is not good, though; it should be simple, high-quality, clean, and only the necessary code. PRs are welcome.

**Reply by datend3nker** [reply]
For now just meant as proof of concept. Once I am happy with the flow and the feature set, ill make my hands dirty.
PR before rewrite or after satisfiable state reached?

**Reply by biodranik** [reply]
After is better => less efforts to review )

**Reply by alexnagorn** [reply]
Is it possible to add one screen more with real map, arrow and built way? Just like on phone. Very useful for tracking and running
<img width="471" height="471" alt="CircleMap_v1" src="https://github.com/user-attachments/assets/fa0a8be3-2cd2-4094-9331-d714c8cd3076" />

**Reply by datend3nker** [reply]
Already on it. But not easy as you would sacrifice power due to rendering or streaming. Newest version has some sort of standalone mode and companion mode. There you'll have a map as well depending on your settings.

Hoping to finish the "vibe" phase this week

**Reply by angelooribeiro** [reply]
Whenever you want to update a beta version I can test it on samsung  ultra and provide feedback.
Would be really nice to have the offline map available in the watch without having to start a route or something. Is power efficient for when you do some marked hike and want to know where is the next water source or hut.

**Comment by datend3nker** [comment]
Sorry for taking so long. Had a lot to do for Uni.
I thing this might be an OKish version to test: [https://github.com/datend3nker/organicmaps-wearos-vibe/releases/tag/v0.2](https://github.com/datend3nker/organicmaps-wearos-vibe/releases/tag/v0.2)
It contains all features I personally would wish for.
Please mind: This is only a POC and entirely written by AI (claude and gemini, architecture was 50% from me).
So this is meant to get an idea of a potential userflow.

Ideas and questions can be discussed here or in the discussions section of the wearos repo

**Reply by biodranik** [reply]
Thanks! Is it possible to record a screen video that demonstrates the app?

**Reply by datend3nker** [reply]
Here is the [video](https://nextcloud.ettner.cloud/s/qJfjJfeAp5KwNF9). It is uncut, so map transfer is a bit long. Also depending on your connection speed, mapstreaming can be much slower. I am currently testing it on my galaxywatch7 and it takes up to 20s till one area is loaded. Still i thing this is a good option if space is tight. To improve map performance, just download or pull map from phone and switch to local map in the settings.

To gain an overview of all features, please consult the readme.
In the releases under "[Sideload / Testing Builds](https://github.com/datend3nker/organicmaps-wearos-vibe/releases/tag/sideload-latest), you can find sideload ready builds for phone and watch.
Would appreciate feedback

**Reply by biodranik** [reply]
Wow, thanks, that's a full app on the watch! Can the app work on the watch without a connection to a smartphone? If no, there's no need to complicate the watch UI and replicate the phone's UI/UX there. If yes, what are the use cases? Do users really need to see the map on the watch? Is it convenient to use?

In general, this prototype looks unusable/too complex and requires a lot of UI/UX simplifications.

**Reply by datend3nker** [reply]
1. Yes it can, as long as settings are set correctly (standalone mode). Maps still need to be downloaded via internet or by pulling from phone.
2. A use case i personally can think of is simple navigation for example by bike. Currently Omaps calls every crossroad with a slight offset to turn right or left and then directly in the other direction. Very confusing when there are multiple tight roads. A quick look at the map can resolve those confusions. For standalone (personally no usecase), user could still navigate even Phone is dead?! Some users talked about deploying the phone app onto the watch, so standalone would resolve that. In case not needed, simply switch it off in the settings
3. totally agree. Happy for suggestions

**Reply by biodranik** [reply]
For (2), the root cause is wrong directions coming from the routing core, and implementing a watch app to fix such a bug is the wrong approach. Would you please make a separate issue with many detailed examples of such "wrong" directions with OSM links, so we can reproduce and fix them properly in the core?

I propose we start with a simple, polished/easy-to-use companion app (relying on the smartphone) that will show/announce the next directions during walking/cycling navigation (and possibly the car one, for easier testing).

WDYT?

**Reply by datend3nker** [reply]
Yes agreed. Incremental but polished development. 
First contribution would include a (empty) watch app, as some backend in the phone app has to be created or at leased established. Once the phone and watchapp can communicate features are added. 
As of visuals I currently would use a similar look:
<img width="221" height="208" alt="image" src="https://github.com/user-attachments/assets/2b6d17c9-d516-41b8-95c7-4feda84f092c" />
<img width="204" height="210" alt="image" src="https://github.com/user-attachments/assets/4f81ffa7-4e3a-4cac-bd03-7b8ed4c9357c" />
(images are taken from the Video...)
or is that already to much?

**Reply by biodranik** [reply]
1. Will the UI be different for different nav modes?
2. It looks small. Would it be convenient for users with imperfect eyesight?
3. Are there dark/light themes on watches?
4. How are other apps showing similar directions? Can we learn the best from them?

**Reply by datend3nker** [reply]
1. No. Atleast not in the early states. Because of safety, I would not implement car navigation.
2. Some sizing/tuning would be required
3. As far as I know Komoot and Google maps use some dark grey
4. As far as I know there are not many apps. Komoot and GMaps are major. 
The first visual would be nice if close to Komoot I guess:
<img width="380" height="385" alt="image" src="https://github.com/user-attachments/assets/84300b3c-4be8-40a0-ad6d-2739b2670e9b" />
<img width="380" height="385" alt="image" src="https://github.com/user-attachments/assets/b016b6a4-68e7-4b8e-924d-3318f9b83390" />
<img width="763" height="735" alt="image" src="https://github.com/user-attachments/assets/b38603fc-b02a-4861-80be-6eaf2dcccaa9" />


So in general some sizing and mybe thicker font. Does having a light theme any benefits on the watch (except map). Does save power and is okish visible in daylight

**Comment by datend3nker** [comment]
I think, this issue is covered by #12030. I think #3273 can be closed as duplicate

**Reply by biodranik** [reply]
Ok, let's continue in https://github.com/organicmaps/organicmaps/issues/12030
